From 721d2f88f1209aae4fd915f7576e2c5194604fbb Mon Sep 17 00:00:00 2001
From: Fabien Dessenne <fabien.dessenne@st.com>
Date: Thu, 2 Feb 2017 16:48:29 +0100
Subject: [PATCH 26/26] STM: waylandsink: set video surface opaque

Unless the video format has an alpha component (ARGB, ...), set the
video_surface opaque.
In the usual case where the black area_surface has the same size as the
video_surface (eg: run gst-play-1.0 video.mp4), this makes the black
surface totally occluded which makes weston compositor's life easier
since it can ignore that surface.
Also unconditionally set the black area_surface opaque.

Change-Id: Ide4992783c6fd06e8ed1777bb07d3d30fd96cf7c
Signed-off-by: Fabien Dessenne <fabien.dessenne@st.com>
Reviewed-on: https://gerrit.st.com/64520
Reviewed-by: CITOOLS <smet-aci-reviews@lists.codex.cro.st.com>
---
 ext/wayland/wlwindow.c | 23 +++++++++++++++++++++++
 1 file changed, 23 insertions(+)

diff --git a/ext/wayland/wlwindow.c b/ext/wayland/wlwindow.c
index 00d733f..b6bad7f 100644
--- a/ext/wayland/wlwindow.c
+++ b/ext/wayland/wlwindow.c
@@ -286,6 +286,28 @@ gst_wl_window_resize_video_surface (GstWlWindow * window, gboolean commit)
   window->surface_height = res.h;
 }
 
+static void
+gst_wl_window_set_opaque (GstWlWindow * window, const GstVideoInfo * info)
+{
+  struct wl_region *region;
+
+  /* Set area opaque */
+  region = wl_compositor_create_region (window->display->compositor);
+  wl_region_add (region, 0, 0, window->render_rectangle.w,
+      window->render_rectangle.h);
+  wl_surface_set_opaque_region (window->area_surface, region);
+  wl_region_destroy (region);
+
+  if (!GST_VIDEO_INFO_HAS_ALPHA (info)) {
+    /* Set video opaque */
+    region = wl_compositor_create_region (window->display->compositor);
+    wl_region_add (region, 0, 0, window->render_rectangle.w,
+        window->render_rectangle.h);
+    wl_surface_set_opaque_region (window->video_surface, region);
+    wl_region_destroy (region);
+  }
+}
+
 void
 gst_wl_window_render (GstWlWindow * window, GstWlBuffer * buffer,
     const GstVideoInfo * info, guint width, guint height)
@@ -300,6 +322,7 @@ gst_wl_window_render (GstWlWindow * window, GstWlBuffer * buffer,
 
     wl_subsurface_set_sync (window->video_subsurface);
     gst_wl_window_resize_video_surface (window, FALSE);
+    gst_wl_window_set_opaque (window, info);
   }
 
   if (G_LIKELY (buffer))
-- 
2.7.4

